tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "tensorflow/core/framework/attr_value.pb.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/grappler/clusters/cluster.h"
+#include "tensorflow/core/grappler/grappler_item.h"
+#include "tensorflow/core/grappler/mutable_graph_view.h"
+#include "tensorflow/core/grappler/op_types.h"
+#include "tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.h"
+#include "tensorflow/core/grappler/optimizers/data/function_utils.h"
+#include "tensorflow/core/grappler/optimizers/data/graph_utils.h"
+#include "tensorflow/core/grappler/utils.h"
+#include "tensorflow/core/platform/protobuf.h"
+
+namespace tensorflow {
+namespace grappler {
+namespace {
+
+constexpr std::array<const char*, 3> kDataDiscarding = {
+    "ShardDataset", "SkipDataset", "TakeDataset",
+};
+
+constexpr std::array<const char*, 6> kCardinalityPreserving = {
+    "CacheDataset", "CacheDatasetV2", "PrefetchDataset",
+    "MapDataset", "ParallelMapDataset", "ParallelMapDatasetV2",
+};
+
+bool IsDataDiscarding(const NodeDef& node) {
+  for (const auto& data_discarding_op : kDataDiscarding) {
+    if (node.op() == data_discarding_op) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool IsCardinalityPreserving(const NodeDef& node) {
+  for (const auto& cardinality_preserving_op : kCardinalityPreserving) {
+    if (node.op() == cardinality_preserving_op) {
+      return true;
+    }
+  }
+  return false;
+}
+
+}  // namepsace
+
+Status HoistDataDiscardingOps::OptimizeAndCollectStats(Cluster* cluster,
+                                                       const GrapplerItem& item,
+                                                       GraphDef* output,
+                                                       OptimizationStats* stats) {
+  *output = item.graph;
+  MutableGraphView graph(output);
+  bool updated;
+  do {
+    updated = false;
+    for (NodeDef node : graph.graph()->node()) {
```

Comment:
Would it make sense to change the order of the loops to make this more efficient? Have an outer loop for all the nodes and in the inner loop, if the node is a discarding op, traverse the parent pointers as long as cardinality preserving ops are encountered.

Reply:
@jsimsa 
I think the current code behaves basically the same as your suggestion. The outer `do .. while` loop is used in case there are new potential op that can be hoist after a round of optimization. For example, if we have
```
TFRecord -> ParallelMap -> Cache -> Skip -> Take -> Repeat
```
The first  round it may only hoist `skip` and we need a second round to hoist `take`.